### DISCUSSION TITLE
SQLAlchemy + Asyncio + Connection Event + asynpg async only interface: Call an async function within an SQLAlchemy event.

### DISCUSSION THREAD

**Discussion by davidandreoletti** [sqlalchemy/sqlalchemy#7580] [Usage Questions] [answered]
### Describe the use case

**Context**
 - App uses SQLAlchemy 1.4.29 with `asynpg` driver.
 - App wants to use custom types defined as Postgres Composite Type. 

**Problem**
Some database drivers (eg: `asyncpg`) can register user defined Postgres Composite types. However registration of said type with `asyncpg` requires a call to an asynchronous function: 

```
await dbapi_connection.set_type_codec('MyCustomType', encoder, decoder, ...))
```

in this particular use case, registration must happen when the connect is first created/connected. This is typically performed in a callback event. 

```
@event.listens_for(engine.sync_engine.pool, "connect")
def regsiter_custom_types(dbapi_connection, ...):
      await dbapi_connection.set_type_codec('MyCustomType', encoder, decoder, ...))
```

However no async function can be called within the 'sync' version of an event. 

PS: SQLAlchemy's ayncio support relies on greenlets to bridge the gap between SQLAlchemy `async` enabled APIs and SQLAlchemy historical `sync` APIs.  I am not familiar enough with greenlelts to know if :
- the current greenlet can be interrupted
- another greenlet can be spawned to run Asyncio coroutine
- interrupted greenlet is resumed

### Databases / Backends / Drivers targeted

Bakcend: Postgres v14
Driver: asynpg 0.25

### Example Use

```
@event.listens_for(engine.sync_engine.pool, "connect")
def regsiter_custom_types(dbapi_connection, ...):
      run_async(dbapi_connection.set_type_codec, 'MyCustomType', encoder, decoder, ...)
```

### Additional context

None

**Comment by zzzeek** [comment, marked as answer]
this functionality is available in sqlalchemy.util, it's not part of public documentation as we did not expect people to need this, but we can add it

```py

from sqlalchemy.util import await_only
@event.listens_for(engine.sync_engine.pool, "connect")
def register_custom_types(dbapi_connection, ...):
      await_only(dbapi_connection.driver_connection.set_type_codec('MyCustomType', encoder, decoder, ...))
```

note that you pass the completed coroutine, already invoked with the arguments, to await_only.  

it would likely be easier if we added a set_type_codec method to the asyncio-adapted connection so people can just cal this directly.

**Reply by zzzeek** [reply]
@Caselit propose we add and document AsyncAdapt_asyncpg_connection.set_type_codec for this use case?

**Reply by CaselIT** [reply]
it does make sense to add it. 

also without using `await_only` the `dbapi_connection` should have an `await_` method available
```py
@event.listens_for(engine.sync_engine.pool, "connect")
def register_custom_types(dbapi_connection, ...):
      dbapi_connection.await_(dbapi_connection.driver_connection.set_type_codec('MyCustomType', encoder, decoder, ...))
```

**Reply by davidandreoletti** [reply]
@zzzeek  / @CaselIT Thanks you for providing an immediate solution, a sneak peak into possible improvement and prompt a response. :-)

**Reply by zzzeek** [reply]
> it does make sense to add it.
> 
> also without using `await_only` the `dbapi_connection` should have an `await_` method available
> 
> ```python
> @event.listens_for(engine.sync_engine.pool, "connect")
> def register_custom_types(dbapi_connection, ...):
>       dbapi_connection.await_(dbapi_connection.driver_connection.set_type_codec('MyCustomType', encoder, decoder, ...))
> ```

we can be consistent with AsyncXYZ.run_sync and also allow them to not have to know about .driver_connection by adding to all the AsyncAdaptedConnections a run_async method:

```python
@event.listens_for(engine.sync_engine.pool, "connect")
def register_custom_types(dbapi_connection, ...):
      dbapi_connection.run_async(lambda connection: connection.set_type_codec('MyCustomType', encoder, decoder, ...))
```

above, the fn returns an awaitable and run_async will await it.

**Reply by CaselIT** [reply]
sure. I believe we can just implement it once in `AdaptedConnection` that's the base for all async connections

**Reply by zzzeek** [reply]
yeah, it can just be on the base
